doc/sphinx/proof-engine/ltac2.rst

```diff
@@ -659,168 +734,476 @@ insert in a concise way an Ltac2 variable of type :n:`constr` into a Coq term.
+      (Tactics that generate integer, string or syntactic values, such as :tacn:`fresh`,
+      are processed during the evaluation phase.)
+
+      Unlike other tactics, `*match*!` tactics get their first success (applying tactics to do
+      so) as part of the evaluation phase.  Among other things, this can affect how early
+      failures are processed in :tacn:`assert_fails`.  Please see the note in :tacn:`assert_fails`.
+
+Variables from the :n:`@tac2pat1` are statically bound in the body of the branch, to
 values of type `constr` for the variables from the :n:`@term` pattern and to a
 value of type `Pattern.context` for the variable :n:`@lident`.
 
 Note that unlike Ltac, only lowercase identifiers are valid as Ltac2
 bindings, so that there will be a syntax error if one of the bound variables
 starts with an uppercase character.
 
-The semantics of this construction is otherwise the same as the corresponding
+The semantics of this construction are otherwise the same as the corresponding
 one from Ltac1, except that it requires the goal to be focused.
 
+.. _ltac2_match_vs_lazymatch_ex:
+
+.. example:: Ltac2 Comparison of lazy_match! and match!
+
+   (Equivalent to this :ref:`Ltac1 example<match_vs_lazymatch_ex>`.)
+
+   These lines define a `msg` tactic that's used in several examples as a more-succinct
+   alternative to `print (to_string "...")`:
+
+   .. From Ltac2 Require Import Message. *outside* the proof lest it disappear with "abort"
+
+   .. coqtop:: in
+
+      From Ltac2 Require Import Message.
+      Ltac2 msg x := print (of_string x).
+
+   .. coqtop:: none
+
+      Goal True.
+
+   In :tacn:`lazy_match!`, if :token:`ltac2_expr` fails, the :tacn:`lazy_match!` fails;
+   it doesn't look for further matches.  In :tacn:`match!`, if :token:`ltac2_expr` fails
+   in a matching branch, it will try to match on subsequent branches.
+
+   .. coqtop:: all
+
+      Fail lazy_match! constr:(True) with
+      | True => msg "branch 1"; fail
+      | _ => msg "branch 2"
+      end.
+
+      match! constr:(True) with
+      | True => msg "branch 1"; fail
+      | _ => msg "branch 2"
+      end.
+
+.. _ltac2_match_vs_multimatch_ex:
+
+.. example:: Ltac2 Comparison of match! and multi_match!
+
+   (Equivalent to this :ref:`Ltac1 example<match_vs_multimatch_ex>`.)
+
+   :tacn:`match!` tactics are only evaluated once, whereas :tacn:`multi_match!`
+   tactics may be evaluated more than once if the following constructs trigger backtracking:
+
+   .. coqtop:: all
+
+      Fail match! constr:(True)with
+      | True => msg "branch 1"
+      | _ => msg "branch 2"
+      end ;
+      msg "branch A"; fail.
+
+   .. coqtop:: all
+
+      Fail multi_match! constr:(True) with
+      | True => msg "branch 1"
+      | _ => msg "branch 2"
+      end ;
+      msg "branch A"; fail.
+
+.. _ltac2_match_with_holes_ex:
+
+.. todo EXAMPLE DOESN'T WORK: Pattern matching?   Matching a pattern with holes
```

Comment:
As I said above, I believe open terms do not work at all in Ltac2 yet

Reply:
We'll find out.  I hope you understand the example is commented out so it won't appear in the generated HTML.  It's a reminder for the future when open terms do work in Ltac2.

Reply:
```suggestion
.. todo EXAMPLE DOESN'T WORK: Ltac2 does not (yet?) handle pattern variables matching open terms.  Matching a pattern with holes
```